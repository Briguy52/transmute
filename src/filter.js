import curry from "./curry";
import { Iterable, Map, Record } from "immutable";
import { filter } from "./protocols/TransmuteCollection";

filter.implement(Array, (test, arr) => arr.filter(test));

filter.implementInherited(Iterable, (test, iter) => iter.filter(test));

filter.implement(Object, (test, obj) => {
  const result = {};
  const keys = Object.keys(obj);
  const len = keys.length;
  for (let i = 0; i < len; i++) {
    const key = keys[i];
    const val = obj[key];
    if (test(val, key, obj)) {
      result[key] = val;
    }
  }
  return result;
});

filter.implementInherited(Record, (test, rec) => {
  return rec.reduce((acc, val, key) => {
    if (!test(val, key, rec)) {
      return acc;
    }
    return acc.set(key, val);
  }, Map());
});

/**
 * Remove values for which `predicate` returns `false`.
 *
 * @example
 * // returns List [ 2 ]
 * filter(
 *   (n) => n % 2 === 0,
 *   List.of(1, 2, 3)
 * );
 *
 * @example <caption>`Record`s have a fixed set of keys, so filter returns a Map instead.</caption>
 * // returns Map { 'one' => 1, 'three' => 3 }
 * filter(
 *   (n) => n % 2 === 0,
 *   ThreeRecord({one: 1, two: 2, three: 3})
 * );
 *
 * @param {Function} predicate returns `true` if a value should be included.
 * @param {Iterable} subject to filter.
 * @return {Iterable} without values that didn't match `predicate`.
 */
export default curry(filter);
