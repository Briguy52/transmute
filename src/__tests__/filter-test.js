import filter from "../filter";
import { List, Map } from "immutable";

describe("transmute/filter", () => {
  const takeEven = filter(n => n % 2 === 0);

  it("filters Arrays", () => {
    expect(takeEven([1, 2, 3])).toMatchSnapshot();
  });

  it("filters Lists", () => {
    expect(takeEven(List.of(1, 2, 3))).toMatchSnapshot();
  });

  it("filters Maps", () => {
    expect(takeEven(Map.of("one", 1, "two", 2, "three", 3))).toMatchSnapshot();
  });

  it("filters Objects", () => {
    expect(takeEven({ one: 1, two: 2, three: 3 })).toMatchSnapshot();
  });
});
