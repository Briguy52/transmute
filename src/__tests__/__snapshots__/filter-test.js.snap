// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmute/filter filters Arrays 1`] = `
Array [
  2,
]
`;

exports[`transmute/filter filters Lists 1`] = `
Immutable.List [
  2,
]
`;

exports[`transmute/filter filters Maps 1`] = `
Immutable.Map {
  two: 2,
}
`;

exports[`transmute/filter filters Objects 1`] = `
Object {
  "two": 2,
}
`;
